hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -637,6 +433,180 @@ boolean detectAndHandleBackupTimeout(long timeoutMillis) {
         return true;
     }
 
+    private boolean engineActive() {
+        NodeState state = context.node.getState();
+        if (state == NodeState.ACTIVE) {
+            return true;
+        }
+
+        boolean allowed = state == NodeState.PASSIVE && (op instanceof AllowedDuringPassiveState);
+        if (!allowed) {
+            notifyError(new HazelcastInstanceNotActiveException("State: " + state + " Operation: " + op.getClass()));
+            remote = false;
+        }
+        return allowed;
+    }
+
+    private void invoke0(boolean isAsync) {
+        if (invokeCount > 0) {
+            throw new IllegalStateException("This invocation is already in progress");
+        } else if (isActive()) {
+            throw new IllegalStateException(
+                    "Attempt to reuse the same operation in multiple invocations. Operation is " + op);
+        }
+
+        try {
+            setCallTimeout(op, callTimeoutMillis);
+            setCallerAddress(op, context.thisAddress);
+            op.setNodeEngine(context.nodeEngine);
+
+            boolean isAllowed = context.operationExecutor.isInvocationAllowed(op, isAsync);
+            if (!isAllowed && !isMigrationOperation(op)) {
+                throw new IllegalThreadStateException(Thread.currentThread() + " cannot make remote call: " + op);
+            }
+            doInvoke(isAsync);
+        } catch (Exception e) {
+            handleInvocationException(e);
+        }
+    }
+
+
+    @SuppressFBWarnings(value = "VO_VOLATILE_INCREMENT",
+            justification = "We have the guarantee that only a single thread at any given time can change the volatile field")
+    private void doInvoke(boolean isAsync) {
+        if (!engineActive()) {
+            return;
+        }
+
+        invokeCount++;
+
+        // register method assumes this method has run before it is being called so that remote is set correctly
+        if (!initInvocationTarget()) {
+            return;
+        }
+
+        setInvocationTime(op, context.clusterClock.getClusterTime());
+        if (!context.invocationRegistry.register(this)) {
```

Comment:
`register` can also throw a `HazelcastOverloadException`. That case is also unhandled.